### PR TITLE
Session object: Set OS/ABI

### LINKED ISSUE #97535: LLVM target's OS/ABI is lost
When compiling for `x86_64-unknown-hermit`, the resulting binaries should have the OS/ABI (`EI_OSABI`) field in the ELF header set to Standalone (`0xff`) as per the `x86_64-unknown-hermit` LLVM target, which it did before `nightly-2022-04-26`. Now, it is set to "UNIX - System V", which I only tested on Linux.

`git-bisect` identified the introduction of synthetic object files for exported symbols https://github.com/rust-lang/rust/pull/95604/commits/773f533eae25129cea7241b74e54f26ce5eebb62 from https://github.com/rust-lang/rust/pull/95604 to have changed this:

```
773f533eae25129cea7241b74e54f26ce5eebb62 is the first bad commit
commit 773f533eae25129cea7241b74e54f26ce5eebb62
Author: Gary Guo <gary@garyguo.net>
Date:   Sat Apr 2 22:54:51 2022 +0100

    Synthesis object file for `#[used]` and exported symbols

 compiler/rustc_codegen_ssa/src/back/link.rs        | 69 ++++++++++++++++++++++
 compiler/rustc_codegen_ssa/src/back/linker.rs      | 46 +++++++++++++++
 compiler/rustc_codegen_ssa/src/back/metadata.rs    |  2 +-
 .../rustc_codegen_ssa/src/back/symbol_export.rs    | 54 ++++++++++++++---
 compiler/rustc_codegen_ssa/src/base.rs             |  7 +++
 compiler/rustc_codegen_ssa/src/lib.rs              |  2 +
 .../rustc_middle/src/middle/exported_symbols.rs    | 10 ++++
 7 files changed, 182 insertions(+), 8 deletions(-)
```

I did not have a deep dive into the code yet, but I assume the synthetic object file has the host's OS/ABI, which is prevalent over the other OS/ABI.

@nbdd0121, do you have any insight? 

Related: https://github.com/hermitcore/rusty-hermit/issues/300

### PR DESCRIPTION
**Pull request by mkroening**
Closes https://github.com/rust-lang/rust/issues/97535.

This depends on
* https://github.com/gimli-rs/object/pull/438

This adapts LLVM's behavior of [`MCELFObjectTargetWriter::getOSABI`](https://github.com/llvm/llvm-project/blob/8c8a2679a20f621994fa904bcfc68775e7345edc/llvm/include/llvm/MC/MCELFObjectWriter.h#L72-L86).

### CODE REVIEW

File: compiler/rustc_codegen_ssa/src/back/metadata.rs [hunk 1]
```diff
@@ -160,17 +160,31 @@ pub(crate) fn create_object_file(sess: &Session) -> Option<write::Object<'static
...
             // that the `+d` target feature represents whether the double
             // float abi is enabled.
             let e_flags = elf::EF_RISCV_RVC | elf::EF_RISCV_FLOAT_ABI_DOUBLE;
-            file.flags = FileFlags::Elf { e_flags };
+            e_flags
         }
-        _ => {}
+        _ => 0,
+    };
+    let llvm_target = sess.target.llvm_target.as_ref();
```
[the remark this answers is not in the document]
Comment (mkroening) [reply compiler/rustc_codegen_ssa/src/back/metadata.rs]: Thanks for the quick review! Done :+1:

**Comment by mkroening** [comment]
object 0.29.0 has been released and this PR is ready for review. :tada:

**Comment by petrochenkov** [comment]
Thanks!
@bors r+

**Pull request merged**